README
NAME
    Sub::Spec::Runner - Run subroutines

VERSION
    version 0.23

SYNOPSIS
    In YourModule.pm:

     package YourModule;

     use 5.010;
     our %SPEC;

     $SPEC{a} = { deps => {run_sub=>'b'}, ... };
     sub a { my %args = @_; say "a"; [200, "OK"] }

     $SPEC{b} = { deps => {run_sub=>'c'}, ... };
     sub b { my %args = @_; say "b"; [200, "OK"] }

     $SPEC{c} = { deps => {all=>[{run_sub=>'d'}, {run_sub=>'e'}]}, ... };
     sub c { my %args = @_; say "c"; [200, "OK"] }

     $SPEC{d} = { deps => {run_sub=>'e'}, ... };
     sub d { my %args = @_; say "d"; [200, "OK"] }

     $SPEC{e} = { ... };
     sub e { my %args = @_; say "e"; [200, "OK"] }

    In main module:

     use Sub::Spec::Runner;

     my $runner = Sub::Spec::Runner->new(load_modules=>0);
     $runner->add('YourModule::a');
     $runner->run;

    Will output:

     e
     d
     c
     b
     a

DESCRIPTION
    NOTICE: This module and the Sub::Spec standard is deprecated as of Jan
    2012. Rinci is the new specification to replace Sub::Spec, it is about
    95% compatible with Sub::Spec, but corrects a few issues and is more
    generic. "Perinci::*" is the Perl implementation for Rinci and many of
    its modules can handle existing Sub::Spec sub specs.

    This class "runs" a set of subroutines. "Running" a subroutine basically
    means loading the module and calling the subroutine, plus a few other
    stuffs. See run() for more details.

    This module uses Log::Any for logging.

    This module uses Moo for object system.

    This module parses the 'run_sub' deps clause in Sub::Spec spec. It
    allows specifying dependency of a subroutine to another subroutine. It
    accept fully qualified subroutine name as value, or a 2-element array
    containing the subroutine name and its arguments. The depended
    subroutine will be add()-ed to the runner. Example:

     run_sub => 'main::func2'

     run_sub => ['My::Other::func', {arg1=>1, arg2=>2}]

ATTRIBUTES
  common_args => HASHREF
    Arguments to pass to each subroutine. Note that each argument will only
    be passed if the 'args' clause in subroutine spec specifies that the sub
    accepts that argument, or if subroutine doesn't have an 'args' clause.
    Example:

     package Foo;

     our %SPEC;
     $SPEC{sub0} = {};
     sub sub0 { ... }

     $SPEC{sub1} = {args=>{}};
     sub sub1 { ... }

     $SPEC{sub2} = {args=>{foo=>"str"}};
     sub sub2 { ... }

     $SPEC{sub3} = {args=>{bar=>"str"}};
     sub sub2 { ... }

     $SPEC{sub4} = {args=>{foo=>"str", bar=>"str"}};
     sub sub4 { ... }

     package main;
     use Sub::Spec::Runner;

     my $runner = Sub::Spec::Runner->new(common_args => {foo=>1, foo=>2});
     $runner->add("Foo::sub$_") for (1 2 3 4);
     $runner->run;

    Then only sub0 and sub4 will receive the 'foo' and 'bar' args. sub1
    won't receive any arguments, sub2 will only receive 'foo', sub3 will
    only receive 'bar'.

  load_modules => BOOL (default 1)
    Whether to load (require()) modules when required (in add()). You can
    turn this off if you are (or want to make) sure that all the subroutines
    to be run are already loaded.

  stop_on_sub_errors => BOOL (default 1)
    When run()-ning subroutines, whether a non-success return value from a
    subroutine stops the whole run. If turned off, run() will continue to
    the next subroutine. Note that you can override what constitutes a
    success return value by overriding success_res().

  allow_add_same_sub => BOOL (default 0)
    If this attribute is set to 0 (the default), then if you add() the same
    subroutine more than once, even with different argument, then the
    subsequent add will be ignored. For example:

     $runner->add('sub1', {arg1=>1}); # 'sub1' added
     $runner->add('sub1', {arg1=>2}); # not added

    If this attribute is set to 1, then you can add() the same subroutine
    more than once but with different arguments.

     $runner->add('sub1', {arg1=>1}); # 'sub1' added with args: {arg1=>1}
     $runner->add('sub1', {arg1=>2}); # 'sub1' added again with args: {arg1=>2}

  order_before_run => BOOL (default 1)
    Before run() runs the subroutines, it will call order_by_dependencies()
    to reorder the added subroutines according to dependency tree (the
    'sub_run' dependency clause). You can turn off this behavior by setting
    this attribute to false.

    If this attribute is a true but negative value, then the order will be
    reversed.

  undo => BOOL (default undef)
    If set to 0 or 1, then these things will happen: 1) Prior to running,
    all added subroutines will be checked and must have 'undo' or 'reverse'
    feature, or are 'pure' (see the 'features' clause in Sub::Spec for more
    details on specifying features). 2) '-undo_action' and '-undo_data'
    special argument will be given with value 0/1 to each sub supporting
    undo (or '-reverse' 0/1 for subroutines supporting reverse). No special
    argument will be given for pure subroutines.

    In summary: setting to 0 means run normally, but instruct subroutines to
    store undo information to enable undo in the future. Setting to 1 means
    undo. Setting to undef (the default) means disregard undo stuffs.

  undo_data_dir => STRING (default ~/.undo)
    Location to put undo data. See save_item_undo_data() for more details.

  dry_run => BOOL (default 0)
    If set to 0 or 1, then these things will happen: 1) Prior to running,
    all added subroutines will be checked and must have 'dry_run' feature,
    or are 'pure' (see 'features' clause in Sub::Spec for more details on
    specifying features). 2) '-dry_run' special argument will be given with
    value 0/1 to each sub supporting dry run. No special argument will be
    given for pure subroutines.

  last_res => RESULT
    This attribute will be set by run() to be the result of each item after
    running each item.

METHODS
  $runner->get_spec($subname) => SPEC
    Get spec for sub named $subname. Will be called by add(). Can be
    overriden to provide your own specs other than from %SPECS package
    variables.

  $runner->add($subname[, $args])
    Add subroutine to the queue of subroutines to be run. Example:

     $runner->add('Package::subname');
     $runner->add('Package::anothersub', {arg=>1});

    Will first load the module (unless load_modules attribute to false).
    Then attempt to get the sub spec by reading the %SPEC package var (this
    behavior can be changed by overriding get_spec()). Will die if cannot
    get spec.

    After that, it will check dependencies (the 'deps' spec clause) and die
    if some dependencies are unmet. All subroutine names mentioned in
    'run_sub' dependency clause will also be add()-ed automatically,
    recursively.

    If the same subroutine is added twice, then it will only be queued once,
    unless allow_add_same_sub attribute is true, in which case duplicate
    subroutine name can be added as long as the arguments are different.

  $runner->empty()
    Empty all items in the queue. This is the opposite of what add() does.

  $runner->order_by_dependencies()
    Reorder set of subroutines by dependencies. Normally need not be called
    manually since it will be caled by run() prior to running subroutines,
    but if you turn off 'order_before_run' attribute, you'll need to call
    this method explicitly if you want ordering.

    Will return a true value on success, or false if dependencies cannot be
    resolved (e.g. there is circular dependency).

  $runner->todo_items() => ARRAYREF
    Return the items in queue not yet run, in order. Previously run
    subroutines can belong to this list again if repeat()-ed.

  $runner->done_items() => ARRAYREF
    Return the items in queue already run. Never-run subroutines can belong
    to this list too if skip()-ed.

  $runner->run(%opts) => [STATUSCODE, ERRMSG, RESULT]
    Run (call) the queue of subroutines previously added by add().

    First it will check 'undo' attribute. If defined, then all added
    subroutines are required to have undo/reverse/pure feature or otherwise
    run() will immediately return with error 412.

    Then it will check 'dry_run' attribute. If true, then all added
    subroutines are required to have dry_run/pure feature or otherwise run()
    will immediately return with error 412.

    After that it will call order_by_dependencies() to reorder the
    subroutines according to dependency order. This can be turned off via
    setting 'order_before_run' attribute to false.

    After that, it will call pre_run(), which you can override. pre_run()
    must return true, or run() will immediately return with 412 error.

    Then it will call each item in the queue successively. Each subroutine
    will be called with arguments specified in 'common_args' attribute and
    args specified in add(), with one extra special argument, '-runner'
    which is the runner object.

    Before running a subroutine: pre_sub() will be called. It must return
    true, or run() will immediately return with 500 error. If 'undo'
    attribute is set to true, get_item_undo_data() will be called to get
    undo data (see get_item_undo_data() for more details on customizing
    storage of undo data).

    The subroutine being run can see the status/result of other subroutines
    by calling $runner->done($subname), $runner->result($subname). It can
    share data by using $runner->stash(). It can also repeat/skip some
    subroutines by calling $runner->skip(), skip_all(), repeat(),
    repeat_all(), branch_done(). It can jump to other subroutines using
    $runner->jump(). See the respective method documentation for more
    details.

    After running a subroutine: Runner will store the return value of each
    subroutine. Exception from subroutine will be trapped by eval() and upon
    exception return value of subroutine is assumed to be 500.

    If subroutine runs successfully: if 'undo' attribute is defined and
    false (meaning normal run but save undo information),
    save_item_undo_data() will be called to save undo information (see
    save_item_undo_data() for more details on customizing storage of undo
    data). If 'undo' attribute is true, remove_item_undo_data() will be
    called to remove undo information.

    After that, post_sub() will be called. It must return true, or run()
    will immediately return with 500 error.

    If 'stop_on_sub_errors' attribute is set to true (the default), then if
    the subroutine returns a non-success result, run() will immediately exit
    with that result. The meaning of subroutine's success can be changed by
    overriding success_res() (by default, all 2xx and 3xx are considered
    success).

    After all subroutines have been run (or skipped), run() will call
    post_run() which must return true. Otherwise run() will immediately exit
    with 500 status.

    After that, run() will return the summary in RESULT (number of
    subroutines run, skipped, successful, etc). It will return status 200 if
    there are at least one subroutine returning success, or 500 otherwise.

    Options for run(), %opts:

    *   use_last_res => BOOL

        Default is false. If set to true, then instead of returning
        statistics/summary result, run() will return the last item's result.

  $runner->format_subname($subname) => STR
    Can be used to format info log message: "Running XXX ..." when about to
    run a subroutine inside run(). Default is "Running Package::bar ..."
    (just the subname)

  $runner->success_res($res) => BOOL
    By default, all responses with 2xx and 3xx are assumed as a success. You
    can override this.

  $runner->pre_run() => BOOL
    See run() for more details. Can be overridden by subclass.

  $runner->pre_sub($subname, $args) => BOOL
    See run() for more details. Can be overridden by subclass.

  $runner->get_item_undo_data($subname, $args) => $undo_data
    Get item's undo data. If 'undo' attribute is set to true, this method
    will be called by run() before running each item.

  $runner->save_item_undo_data($subname, $args, $undo_data)
    Save item's undo data. If 'undo' attribute is defined, this method will
    be called by run() after each item is called, to save the resulting undo
    data.

    The default implementation saves undo data in YAML files under directory
    specified by 'undo_data_dir', one file per subname, the file being named
    <subname>.yaml ("::" replaced with by "." because it is not valid in
    some filesystems). The data being stored is an array of hashes
    ("records"). Each record contains: {args => ..., undo_datas => [step,
    ...]}. Because the Sub::Spec specification allows different undo data
    for different arguments, the YAML file store undo data for each
    different argument in each record.

    To store undo data in other location, you can set the 'undo_data_dir' or
    alternatively override the *_undo_data() methods and provide your own
    implementation.

  $runner->get_sub_undo_data($subname) => $records
    Get all undo data records for a certain subname. See
    save_item_undo_data() for more details.

  $runner->remove_item_undo_data($subname, $args)
    Remove an item's undo data. If 'undo' attribute is set to true, will be
    called by run() after a successful undo.

  $runner->remove_sub_undo_data($subname, $args)
    Remove a sub's undo data.

  $runner->post_sub($subname, $args) => BOOL
    See run() for more details. Can be overridden by subclass.

  $runner->post_run() => BOOL
    See run() for more details. Can be overridden by subclass.

  $runner->result($subname[, $args]) => RESULT
    Return the result of run subroutine named SUBNAME. If subroutine is not
    run yet, will return undef.

    If there are multiple items matching $subname, only the first item's
    result will be returned.

  $runner->is_done($subname[, $args]) => BOOL
    Check whether an item is already run/done.

  $runner->done(SUBNAME[, ARGS[, VALUE]]) => OLDVAL
    If VALUE is set, set a subroutine to be done/not done. Otherwise will
    return the current done status of SUBNAME.

    SUBNAME can also be a regex, which means all subroutines matching the
    regex. The last SUBNAME's current done status will be returned.

  $runner->skip(SUBNAME[, ARGS])
    Alias for done(SUBNAME, ARGS, 1).

  $runner->skip_all()
    Skip all subroutines.

  $runner->repeat(SUBNAME[, ARGS])
    Alias for done(SUBNAME, ARGS, 0).

  $runner->repeat_all()
    Repeat all subroutines.

  $runner->done_branch(SUBNAME, ARGS, VALUE)
    Just like done(), except that will set SUBNAME *and all its dependants*.
    Example: if a depends on b and b depends on c, then doing
    branch_done('c', undef, 1) will also set a & b as done.

  $runner->skip_branch(SUBNAME[, ARGS])
    Alias for done_branch(SUBNAME, ARGS, 1).

  $runner->repeat_branch(SUBNAME[, ARGS])
    Alias for done_branch(SUBNAME, ARGS, 0).

  $runner->jump($subname, $args)
    Jump to another item. Can be called in pre_sub() or inside subroutine or
    post_sub().

  $runner->stash(NAME[, VALUE]) => OLDVAL
    Get/set stash data. This is a generic place to share data between
    subroutines being run.

FAQ
  What is the point of this module?
    Sub::Spec allows us to add various useful metadata to subroutines, like
    dependencies and specific features. Sub::Spec::Runner utilizes this
    information to make calling subroutines a bit more like running
    programs/installing software packages, e.g.:

    *   checking requirements prior to calling a subroutine;

        For example, you can specify that backup_db() requires the program
        "/usr/bin/mysqldump".

    *   reordering list of subroutines to run according to
        interdependencies;

        You can specify that a() and b() must be run before c(), d() must be
        run before c(), and so on. The runner will automatically resolve
        dependencies by loading required modules and reorder subroutine
        execution.

    *   running in dry-run mode;

        See 'dry_run' attribute for more details.

    *   running in undo mode;

        See 'undo' attribute for mor details.

    *   summary/statistics;

        Including the number of subroutines run, number of
        successes/failures, etc.

  What are some of the applications for this module?
    Sub::Spec::CmdLine uses this module in a straightforward way.
    Sub::Spec::CmdLine allows you to run subroutines from the command-line.

    Our Spanel project uses this module to run "setuplets", which are
    hosting server setup routines broken down to smaller bits. Each bit can
    be run individually (in dry-run and/or undo mode) but dependencies will
    always be respected (e.g. setuplets A requires running B, so running
    only A will always run B first).

SEE ALSO
    Sub::Spec

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Sub::Spec::Runner

0.23     2012-01-20

         No functional changes. Mark deprecated.


0.22     2011-10-23

         No functional changes. Replace Data::Sah::Util with Data::Sah.


0.21     2011-10-22

         No functional changes. Replace old Sub::Spec::Clause::* with
         Sub::Spec::DepChecker.


0.20     2011-10-22

         No functional changes. Updated because _parse_schema() is moved from
         Sub::Spec::Utils to Data::Sah::Util and 'deps' clause handling is moved
         to Sub::Spec::DepChecker.


0.19     2011-06-02

         [INCOMPATIBLE CHANGES]

         * {get,save,remove}_undo_data() becomes
           {get,save,remove}_item_undo_data().

         [NEW FEATURES]

         * Add methods: {get,remove}_sub_undo_data().


0.18     2011-06-01

         [INCOMPATIBLE CHANGES]

         * run() now always returns a summary status (e.g. 500 if all items
           failed, 200 if some succeeded, etc) instead of last item's result.
           See run option 'use_last_res' to return last item's result instead.

         [NEW FEATURES]

         * run() now accepts options. Currently known option: use_last_res.

         * last_res() for easier access to last item's result.


0.17     2011-05-31

         [NEW FEATURES]

         * New method: empty().


0.16     2011-05-31

         [INCOMPATIBLE CHANGES]

         * Setting 'undo' attribute no longer automatically reverse order of
           run. To reverse order, set 'order_before_run' to a true but negative
           value.


0.15     2011-05-26

         This release introduces a new feature: under a new flag attribute
         'allow_add_same_sub', a single subroutine name can be add()-ed multiple
         times, as long as the arguments differ. There are some incompatible
         changes introduced as well, as a consequence.

         [INCOMPATIBLE CHANGES]

         * run() returns 412 instead of 400 when there are no items added.

         * Argument list for done(), skip(), repeat(), jump(), pre_sub(),
           post_sub() changes, due to extra argument $args when specifying
           items.

         * rename methods: todo_subs() -> todo_items() (and return hash data
           instead of subname), done_subs() -> done_items() (ditto).

         * branch_done() replaced by skip_branch() and repeat_branch().

         [NEW FEATURES]

         * New flag attribute 'allow_add_same_sub' to allow a single subroutine
           name to be add()-ed multiple times as long as the arguments differ.

         * 'run_sub' dep clause can accept [SUBNAME, ARGS] instead of just
           SUBNAME.

         [ENHANCEMENTS]

         * run() now returns these keys also in data: num_items,
           num_success_items, num_failed_items, num_skipped_items to
           differentiate with num*subs and num*runs keys.

         * add is_done()

0.14     2011-05-16

         [INCOMPATIBLE CHANGES]

         * Update to latest Sub::Spec specification: -undo becomes -undo_action,
           -state becomes -undo_data. Saves undo data in ~/.subspec/.undo by
           default.


0.13     2011-03-11

         [ENHANCEMENTS]

         * Support 'result_naked' spec clause.


0.12     2011-03-10

         [DOCUMENTATION]

         * Add FAQ.


0.11     2011-03-10

         [BUG FIXES]

         * Fix small bug in SS::Runner::State's delete().


0.10     2011-03-10

         [ENHANCEMENTS]

         * Add 'undo' and 'dry_run' attributes.


0.09     2011-03-03

         [INCOMPATIBLE CHANGES]

         * 'args' attribute is now 'common_args'.

         [ENHANCEMENTS]

         * Add($subname[, $args]): accepts per-sub args.


0.08     2011-02-28

         [INCOMPATIBLE CHANGES]

         * Rename module: Sub::Spec::Runner::Orderly -> Sub::Spec::Runner.

         * Update to latest Sub::Spec: 'depends' clause becomes 'dep'. 'sub' dep
           clause becomes 'run_sub'.

         * Check dependencies (using SS::Clause::deps::check()) before add()-ing.

         * run() no longer accepts %opts, use the new attributes:
           stop_on_sub_errors and order_before_run instead.

         [ENHANCEMENTS]

         * Add result().


0.07     2011-02-25

         Add missing dep [CT].


0.06     2011-02-24

         Add get_spec() to allow overriding behaviour for getting sub specs,
         remove $spec argument to add().

         Add option order_by_dependencies to run() to allow run() to not run
         order_by_dependencies().

         Some POD fixes.


0.05     2011-02-24

         Update doc.


0.04     2011-02-24

         Rewrite into OO-style, rename from Sub::Spec::RunDeps ->
         Sub::Spec::Runner::Orderly.


0.03     2011-02-23

         Add missing dep: Log::Any.


0.02     2011-02-22

         Add before_item & after_item argument.


0.01     2011-02-22

         First release.


lib/Sub/Spec/Runner/State.pm
package Sub::Spec::Runner::State;
BEGIN {
  $Sub::Spec::Runner::State::VERSION = '0.13';
}
# ABSTRACT: Save undo data

use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use Moo;

use File::Path qw(make_path);
use YAML::Syck qw(LoadFile DumpFile);

has root_dir => (is => 'rw', default=>sub{"$ENV{HOME}/.subspec/.undo"});

# By default undo data is saved in ~/.subspec/undo/<DOTTED_SUBNAME>.yaml or
# <DOTTED_SUBNAME>/, e.g. ~/.subspec/.undo/perl.yaml


# get([\%opts], key)
sub get {
    my $self = shift;
    my $opts;
    if (ref($_[0]) eq 'HASH') {
        $opts = shift;
    } else {
        $opts = {};
    }
    $self->{subname} = $opts->{subname} // (caller(1))[3];

    $self->_load;
    $self->{data}{$_[0]};
}

# set([\%opts], key=>val, key2=>val, ...)
sub set {
    my $self = shift;
    my $opts;
    if (ref($_[0]) eq 'HASH') {
        $opts = shift;
    } else {
        $opts = {};
    }
    $self->{subname} = $opts->{subname} // (caller(1))[3];

    use autodie;

    $self->_load;
    while (my ($key, $val) = splice @_, 0, 2) {
        $self->{data}{$key} = $val;
        $log->tracef("Setting key: %s => %s", $key, $val);
    }
    $log->tracef("Writing file %s (%d keys) ...",
                 $self->{path_file}, scalar(keys %{$self->{data}}));
    DumpFile($self->{path_file}, $self->{data});
}

# delete([\%opts], key, key2, ...)
sub delete {
    my $self = shift;
    my $opts;
    if (ref($_[0]) eq 'HASH') {
        $opts = shift;
    } else {
        $opts = {};
    }
    $self->{subname} = $opts->{subname} // (caller(1))[3];

    use autodie;

    $self->_load;
    delete $self->{data}{$_} for @_;
    if (scalar keys %{$self->{data}}) {
        $log->tracef("Writing file %s (%d keys) ...",
                     $self->{path_file}, scalar(keys %{$self->{data}}));
        DumpFile($self->{path_file}, $self->{data});
    } else {
        $log->tracef("Deleting file %s (no more keys) ...",
                     $self->{path_file});
        unlink $self->{path_file} if -f $self->{path_file};
    }
}

sub _load {
    my ($self) = @_;

    use autodie;

    $self->_prepare_storage();
    unless (-f $self->{path_file}) {
        $self->{data} = {};
        return;
    }
    my $mtime = (-M $self->{path_file});
    if (!$self->{data} ||
            !$self->{path_file_mtime} ||
                $self->{path_file_mtime} != $mtime) {
        $log->trace("Reading file $self->{path_file} ...");
        $self->{data} = LoadFile($self->{path_file});
        $self->{path_file_mtime} = $mtime;
    }
}

sub _prepare_storage {
    my ($self, $opts) = @_;

    my $file = $self->{subname};
    $file =~ s/^Spanel::Setup:://;
    $file =~ s/::/./g;

    use autodie;

    make_path($self->{root_dir}, {mode=>0700}) unless (-d $self->{root_dir});
    $self->{path_file} = "$self->{root_dir}/$file.yaml";
    $self->{path_dir}  = "$self->{root_dir}/$file";
}

1;

__END__
=pod

=head1 NAME

Sub::Spec::Runner::State - Save undo data

=head1 VERSION

version 0.13

=for Pod::Coverage get set delete

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Run subroutines",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Sub-Spec-Runner",
  "dist_version" => "0.23",
  "license" => "perl",
  "module_name" => "Sub::Spec::Runner",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Algorithm::Dependency" => 0,
    "Array::Functions::Undoable" => 0,
    "Capture::Tiny" => 0,
    "Data::Sah" => 0,
    "File::chdir" => 0,
    "JSON" => 0,
    "Log::Any" => 0,
    "Moo" => 0,
    "Sub::Spec" => "0.08",
    "Sub::Spec::DepChecker" => 0,
    "Test::More" => "0.96",
    "YAML::Syck" => 0,
    "perl" => "5.010000"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
